mm/kfence.c

```diff
@@ -516,12 +543,89 @@ bool kfence_handle_page_fault(unsigned long addr)
 	return kfence_unprotect(addr);
 }
 
+void kfence_cache_register(struct kmem_cache *s)
+{
+	unsigned long flags;
+	int index;
+	const char *name;
+
+	if (!s)
+		return;
+
+	if (!s->name)
+		name = "ANON";
+	else
+		name = s->name;
+
+	if (s->size > PAGE_SIZE) {
+		pr_debug("skipping cache %s because of size: %d\n", name,
+			 s->size);
+		return;
+	}
+	if (s->ctor) {
+		pr_debug("skipping cache %s because of ctor\n", name);
+		return;
+	}
+	if (s->flags & SLAB_TYPESAFE_BY_RCU) {
+		pr_debug("skipping cache %s because of RCU\n", name);
+		return;
+	}
+	pr_debug("registering cache %s\n", name);
+	spin_lock_irqsave(&kfence_caches_lock, flags);
+	if (kfence_num_caches == KFENCE_MAX_CACHES)
+		goto leave;
+	index = find_cache(s);
+	if (index == -1) {
+		kfence_registered_caches[kfence_num_caches - 1] = s;
+		kfence_num_caches++;
+	}
+leave:
+	spin_unlock_irqrestore(&kfence_caches_lock, flags);
+}
+EXPORT_SYMBOL(kfence_cache_register);
+
+/*
+ * TODO: tear down objects from the deleted cache. We may want to store a
+ * bitmask of KFENCE objects for every registered cache.
+ */
+void kfence_cache_unregister(struct kmem_cache *s)
+{
+	unsigned long flags;
+	int index;
+
+	if (!s)
+		return;
+
+	spin_lock_irqsave(&kfence_caches_lock, flags);
+	if (kfence_num_caches == 0)
+		goto leave;
+	index = find_cache(s);
+	if (index == -1)
+		goto leave;
+	if (index == kfence_num_caches - 1)
+		kfence_registered_caches[index] = NULL;
```

Comment:
Why is the setting to NULL necessary, if you do a kfence_num_caches-- below?

There are 2 options:
1. Either this stays, and in the else-case kfence_registered_caches[kfence_num_caches-1] is also set to NULL.
2. Or this is just removed.

Unless I missed something?

Reply:
True. The initial implementation allowed gaps in the array, this is left over.
I'll remove the redundant store.